platforms/common/include/px4_platform_common/events.h

```diff
@@ -0,0 +1,131 @@
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#pragma once
+
+// Console event output can be enabled if needed.
+// Note: this will only print custom (PX4) events, not from common.
+// Helpful for debugging and to ensure the system is not spammed with events.
+// It does not print arguments.
+#if 1
+#include <px4_platform_common/log.h>
+#define CONSOLE_PRINT_EVENT(log_level, id, str) PX4_INFO_RAW("Event 0x%08x: %s\n", id, str)
+#else
+#define CONSOLE_PRINT_EVENT(log_level, id, str)
+#endif
+
+#include <events/events_generated.h>
+#include <libevents_definitions.h>
+
+#include <stdint.h>
+
+namespace events
+{
+
+namespace util
+{
+
+// source: https://gist.github.com/ruby0x1/81308642d0325fd386237cfa3b44785c
+constexpr uint32_t val_32_const = 0x811c9dc5;
+constexpr uint32_t prime_32_const = 0x1000193;
+constexpr uint64_t val_64_const = 0xcbf29ce484222325;
+constexpr uint64_t prime_64_const = 0x100000001b3;
+inline constexpr uint32_t hash_32_fnv1a_const(const char *const str, const uint32_t value = val_32_const) noexcept
+{
+	return (str[0] == '\0') ? value : hash_32_fnv1a_const(&str[1], (value ^ uint32_t(str[0])) * prime_32_const);
+}
+
+template<typename T>
+inline constexpr void fillEventArguments(uint8_t *buf, T arg)
+{
+	// This assumes we're on little-endian
+	memcpy(buf, &arg, sizeof(T));
+}
+
+template<typename T, typename... Args>
+inline constexpr void fillEventArguments(uint8_t *buf, T arg, Args... args)
+{
+	fillEventArguments(buf, arg);
+	fillEventArguments(buf + sizeof(T), args...);
+}
+
+constexpr unsigned sizeofArguments() { return 0; }
+
+template <typename T, typename... Args>
+constexpr unsigned sizeofArguments(const T &t, const Args &... args)
+{
+	return sizeof(T) + sizeofArguments(args...);
+}
+
+} // namespace util
+
+
+/**
+ * publish/send an event
+ */
+void send(EventType &event);
+
+/**
+ * Generate event ID from an event name
+ */
+constexpr uint32_t ID(const char *name)
```

Comment:
Make sure this can only bind to string literals

```suggestion
template<size_t N>
constexpr uint32_t ID(const char (&name)[N])
```

Reply:
Updated, thanks